Remove zoom on font-relative units when resolving SVGLengthValue

nullhook opened the pull request:
#### ea6f87ed951e23583099f071feac8f878454f948
<pre>
Remove zoom on font-relative units when resolving SVGLengthValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=298102">https://bugs.webkit.org/show_bug.cgi?id=298102</a>
<a href="https://rdar.apple.com/159448970">rdar://159448970</a>

Reviewed by Tim Nguyen.

Currently, when resolving SVGLengthValue, somewhere in the callstack
we call computeUnzoomedNonCalcLengthDouble to get user units as the
final value. This value can sometimes be resolved against the computed
font size, which may or may not include zoom when units are
font-relative or root font-relative.

In SVG contexts when using font-relative units, we expect these values
to be unzoomed because SVG applies zoom during layout-time
at viewport level. (see RenderSVGViewportContainer)

Ideally, a routine named &quot;computeUnzoomed*&quot; should guarantee
non-zoomed values, but this isn&apos;t the case.

This PR removes the applied zoom for font-relative or root
font-relative units in SVG length context.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-font-relative-units-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-font-relative-units-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-font-relative-units.html: Added.
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::computeNonCalcLength const):
(WebCore::SVGLengthContext::removeZoomFromFontOrRootFontRelativeLength const):
* Source/WebCore/svg/SVGLengthContext.h:

Canonical link: <a href="https://commits.webkit.org/299637@main">https://commits.webkit.org/299637@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8d0a7ecff5982195ce844f350bbb160efa29601

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119644 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39338 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29990 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125931 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71718 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ca781931-2732-4a8f-9dc5-967d3e3a38a6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121520 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47916 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90874 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ccb5dc96-a9e9-4fd4-9a92-7e71803c1200) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31958 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107280 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71389 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/04d68731-b8cf-47f4-a6c2-7ea38ca5c517) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25385 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69576 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101426 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25577 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128892 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46566 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35280 "Found 1 new test failure: http/tests/cache/disk-cache/disk-cache-302-status-code.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99473 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46931 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99314 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25232 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44744 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22764 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43130 "Hash e8d0a7ec for PR 50067 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46428 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52134 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45894 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49243 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47580 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6f3cd1cc28146391563138d419895b114006b492. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50067

nikolaszimmermann commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -272,7 +272,23 @@ float SVGLengthContext::computeNonCalcLength(float inputValue, CSS::LengthUnit u
+                : conversionData->fontCascadeForFontUnits();
+            auto& fontDescription = fontCascade.fontDescription();
+
+            float computedSize = fontDescription.computedSize();
```
Is this guaranteed to be unclamped? I vaguely recall some code to clamp the font size above a minimum value.

Alternatively, you could directly access the effective zoom, no?

nullhook replied:
> Is this guaranteed to be unclamped? I vaguely recall some code to clamp the font size above a minimum value.

```
fontDescription.setComputedSize(Style::computedFontSizeFromSpecifiedSize(size, fontDescription.isAbsoluteSize(), useSVGZoomRules(), &style(), document()));
```

are you referring to `computedFontSizeFromSpecifiedSize`? at style-resolution time

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2fa4a48e51ef45850cf09fb9fb84e2b7521ef789. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50067

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/617079cad1645ee66173049fade9e71930d42da6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50067

nt1m commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -268,14 +268,37 @@ ExceptionOr<float> SVGLengthContext::convertValueFromUserUnits(float value, SVGL
+
+    auto resolvedValue = clampTo<float>(Style::computeNonCalcLengthDouble(inputValue, unit, *conversionData));
+
+    if (unit == CSS::LengthUnit::Em || unit == CSS::LengthUnit::Rem) {
```
Can we have test cases for other font units as well?

It would also good to add a comment why the other font units don't have special logic (maybe FIXME).

nullhook replied:
added comment

nt1m reviewed (approved):
This probably could be cleaned up instead of doing the dance of adding zoom and removing zoom but it's probably ok for now.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a944e1e1b57146d56a4918c83e44dd6019bf5304. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50067

nt1m reviewed (commented):
Can we add more testcases to cover other units

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8d0a7ecff5982195ce844f350bbb160efa29601. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50067

nullhook committed:
Remove zoom on font-relative units when resolving SVGLengthValue
https://bugs.webkit.org/show_bug.cgi?id=298102
rdar://159448970

Reviewed by Tim Nguyen.

Currently, when resolving SVGLengthValue, somewhere in the callstack
we call computeUnzoomedNonCalcLengthDouble to get user units as the
final value. This value can sometimes be resolved against the computed
font size, which may or may not include zoom when units are
font-relative or root font-relative.

In SVG contexts when using font-relative units, we expect these values
to be unzoomed because SVG applies zoom during layout-time
at viewport level. (see RenderSVGViewportContainer)

Ideally, a routine named "computeUnzoomed*" should guarantee
non-zoomed values, but this isn't the case.

This PR removes the applied zoom for font-relative or root
font-relative units in SVG length context.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-font-relative-units-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-font-relative-units-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-font-relative-units.html: Added.
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::computeNonCalcLength const):
(WebCore::SVGLengthContext::removeZoomFromFontOrRootFontRelativeLength const):
* Source/WebCore/svg/SVGLengthContext.h:

Canonical link: https://commits.webkit.org/299637@main
